AsyncDisplayKit/Details/ASCollectionDataController.mm

```diff
@@ -15,48 +15,67 @@
 #import "ASCellNode.h"
 #import "ASDataController+Subclasses.h"
 #import "ASIndexedNodeContext.h"
+#import "ASSectionContext.h"
 
 //#define LOG(...) NSLog(__VA_ARGS__)
 #define LOG(...)
 
 @interface ASCollectionDataController () {
   BOOL _dataSourceImplementsSupplementaryNodeBlockOfKindAtIndexPath;
+  BOOL _dataSourceSupportsSectionContext;
+  NSMutableArray<id<ASSectionContext>> *_sectionContexts;
+  NSArray<id<ASSectionContext>> *_pendingSectionContexts;
 }
 
 - (id<ASCollectionDataControllerSource>)collectionDataSource;
 
 @end
 
 @implementation ASCollectionDataController {
-  NSMutableDictionary<NSString *, NSMutableArray<ASIndexedNodeContext *> *> *_pendingContexts;
+  NSMutableDictionary<NSString *, NSMutableArray<ASIndexedNodeContext *> *> *_pendingNodeContexts;
 }
 
 - (instancetype)initWithDataSource:(id<ASCollectionDataControllerSource>)dataSource
 {
   self = [super initWithDataSource:dataSource];
   if (self != nil) {
-    _pendingContexts = [NSMutableDictionary dictionary];
+    _pendingNodeContexts = [NSMutableDictionary dictionary];
     _dataSourceImplementsSupplementaryNodeBlockOfKindAtIndexPath = [dataSource respondsToSelector:@selector(dataController:supplementaryNodeBlockOfKind:atIndexPath:)];
+    _dataSourceSupportsSectionContext = [dataSource supportsSectionContextInDataController:self];
 
     ASDisplayNodeAssertTrue(_dataSourceImplementsSupplementaryNodeBlockOfKindAtIndexPath || [dataSource respondsToSelector:@selector(dataController:supplementaryNodeOfKind:atIndexPath:)]);
+    
+    if (_dataSourceSupportsSectionContext) {
+      _sectionContexts = [NSMutableArray array];
+    }
   }
   return self;
 }
 
 - (void)prepareForReloadDataWithSectionCount:(NSInteger)newSectionCount
 {
   NSIndexSet *sections = [NSIndexSet indexSetWithIndexesInRange:NSMakeRange(0, newSectionCount)];
+  
+  if (_dataSourceSupportsSectionContext) {
+    [_sectionContexts removeAllObjects];
+    [self _populateSectionContextsFromDataSourceWithSectionIndexSet:sections];
+  }
+  
   for (NSString *kind in [self supplementaryKinds]) {
     LOG(@"Populating elements of kind: %@", kind);
     NSMutableArray<ASIndexedNodeContext *> *contexts = [NSMutableArray array];
     [self _populateSupplementaryNodesOfKind:kind withSections:sections mutableContexts:contexts];
-    _pendingContexts[kind] = contexts;
+    _pendingNodeContexts[kind] = contexts;
   }
 }
 
 - (void)willReloadDataWithSectionCount:(NSInteger)newSectionCount
 {
-  [_pendingContexts enumerateKeysAndObjectsUsingBlock:^(NSString * _Nonnull kind, NSMutableArray<ASIndexedNodeContext *> * _Nonnull contexts, __unused BOOL * _Nonnull stop) {
+  NSIndexSet *sectionIndexSet = [NSIndexSet indexSetWithIndexesInRange:NSMakeRange(0, newSectionCount)];
+  
+  [self applyPendingSectionContextsAtIndexSet:sectionIndexSet];
```

Comment:
`applyPendingSectionContextsAtIndexes` and `_populateSectionContextsFromDataSourceForSections`